### DISCUSSION TITLE
Export template output returns newline

### DISCUSSION THREAD

**Discussion by iamjpy** [netbox-community/netbox#17280] [General]
I'm using jinja2 export template within my custom scripts. When outputting the rendered result, some lines break in the middle and return to a newline

Example output from custom script rendering data with an export template:
```
- NAT IP:  192.168.x.x
- Host IP: 10.11.12.13

- NAT IP:  192.168.x.y
- Host IP: 
10.14.15.16

- NAT IP:  192.168.x.z
- Host IP: 
10.17.18.19

- NAT IP: 192.168.x.a
- Host IP: 
10.20.21.22
```
The first two lines are the desired output, then the following entries break after 'Host IP:' and return the string to a new line which is not the expected behavior

Here is a snippet of the export template:
```
{% for ip in queryset.dnat_ip_list %}
- NAT IP: {{ip[0]}}
- Host IP: {{ip[1]}}
{% endfor %}
```

Here is an example input:
```
dnat_ip_list =[['192.168.x.x', '10.11.12.13'], ['192.168.x.y', '10.14.15.16'], ['192.168.x.z', '10.17.18.19'], ['192.168.x.a', '10.20.21.22']]
```

That was a simple example to share, but I have a couple of other examples that result in broken device commands. This make it challenging to push it to a network device if in an array of device commands, the broken command is submitted as two lines which make it invalid instead of the full command as one element

Any help is appreciated!

**Comment by williammarget** [comment]
It looks like you're encountering issues with line breaks in the output of your Jinja2 template rendering, particularly with how the rendered result is being split across lines.

Here are a few things you can check and try to resolve this issue:

### 1. **Check for Unintentional Line Breaks**

Ensure there are no unintended line breaks in your Jinja2 template or input data. You can use a tool or script to visualize hidden characters or whitespace in your text.

### 2. **Whitespace Control in Jinja2**

Jinja2 provides control over whitespace handling which might help in your case. Try using the `-` character in the template to control whitespace:

```jinja2
{% for ip in queryset.dnat_ip_list %}
- NAT IP: {{ ip[0] }}
- Host IP: {{ ip[1] -}}
{% endfor %}
```

Here, `-` after `{{ ip[1] }}` will strip any trailing newline characters.

### 3. **Preprocess Input Data**

If your data has extra whitespace or newline characters, preprocess it before rendering. You can strip unwanted spaces in your script before passing data to the template.

For example:

```python
dnat_ip_list = [[ip[0].strip(), ip[1].strip()] for ip in dnat_ip_list]
```

### 4. **Post-process Output**

If controlling whitespace in the template doesn’t work, you can post-process the output in your script. For instance, you can use regular expressions or string manipulation to remove unwanted line breaks:

```python
import re

# Assuming `rendered_output` is the output from your Jinja2 rendering
cleaned_output = re.sub(r'(?<=Host IP:\s*)\n', '', rendered_output)
```

This regex removes newline characters that follow "Host IP:".

### 5. **Ensure Proper Output Handling**

Make sure that when you handle the rendered output, it doesn’t get split by some other part of your processing pipeline. If you’re writing to files or sending to devices, ensure that there’s no additional processing or formatting that might cause line breaks.

### Example Debugging Steps

You might also want to print out the raw rendered output to see exactly what is being produced:

```python
print(repr(rendered_output))
```

The `repr()` function will show special characters like `\n` explicitly, which can help in diagnosing where the line breaks are coming from.

### Example Jinja2 Template and Output

With these techniques, the output should look like:

```
- NAT IP: 192.168.x.x
- Host IP: 10.11.12.13

- NAT IP: 192.168.x.y
- Host IP: 10.14.15.16

- NAT IP: 192.168.x.z
- Host IP: 10.17.18.19

- NAT IP: 192.168.x.a
- Host IP: 10.20.21.22
```

Each entry is consistent and no extra line breaks should appear.

**Comment by iamjpy** [comment]
You nailed it! There were some hidden `\r\n` in my data so I used `strip()` to preprocess my inputs. Thanks for the assist
